File: loopctl/mcp_runtime.py
#!/usr/bin/env python3
"""Default-deny stateless MCP runtime for loopctl.

Every call resolves one immutable Git subject, validates the policy and selected
module at that subject, checks out a disposable worktree, removes tracked files
outside the selected module's transitive capability closure, executes only the
contract-derived argv/carrier, bounds output, and removes the worktree.

Transport: line-delimited JSON-RPC 2.0 over stdio. Supported methods are
initialize, tools/list, and tools/call.
"""

from __future__ import annotations

import base64
from contextlib import contextmanager
import hashlib
import json
import os
from pathlib import Path
import re
import shutil
import subprocess
import sys
import tempfile
from typing import Any, Iterator

HERE = Path(__file__).resolve().parent
PROTOCOL = "2024-11-05"
SHA40 = re.compile(r"^[0-9a-f]{40}$")
TAG = re.compile(r"^v[0-9][0-9A-Za-z._-]*$")
SAFE_ARTIFACT = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._/-]*$")
SECRET_NAMES = {
    "ANTHROPIC_API_KEY",
    "CLAUDE_CODE_OAUTH_TOKEN",
    "CODEX_ACCESS_TOKEN",
    "CODEX_API_KEY",
    "E2B_API_KEY",
    "FORGEJO_PASSWORD",
    "FORGEJO_TOKEN",
    "GEMINI_API_KEY",
    "GITHUB_TOKEN",
    "OPENAI_API_KEY",
}


class McpError(ValueError):
    """An MCP policy, request, or immutable execution contract is invalid."""


def canonical(value: Any) -> bytes:
    return json.dumps(
        value, ensure_ascii=False, sort_keys=True, separators=(",", ":")
    ).encode("utf-8")


def digest_value(value: Any) -> str:
    return hashlib.sha256(canonical(value)).hexdigest()


def repo_root() -> Path:
    process = subprocess.run(
        ["git", "-C", str(HERE), "rev-parse", "--show-toplevel"],
        capture_output=True,
        text=True,
        check=False,
    )
    if process.returncode != 0:
        raise McpError(process.stderr.strip() or "not inside a Git worktree")
    return Path(process.stdout.strip())


def git(root: Path, *args: str) -> str:
    process = subprocess.run(
        ["git", "-C", str(root), *args],
        capture_output=True,
        text=True,
        check=False,
    )
    if process.returncode != 0:
        raise McpError(process.stderr.strip() or f"git {' '.join(args)} failed")
    return process.stdout.strip()


def resolve_ref(root: Path, ref: str) -> tuple[str, str]:
    if ref in {"HEAD", "main", "master", "trunk"}:
        raise McpError(f"mutable ref is refused: {ref}")
    if not (SHA40.fullmatch(ref) or TAG.fullmatch(ref)):
        raise McpError("ref must be an exact 40-hex commit or immutable v* tag")
    commit = git(root, "rev-parse", f"{ref}^{{commit}}")
    tree = git(root, "rev-parse", f"{commit}^{{tree}}")
    if not SHA40.fullmatch(commit) or not SHA40.fullmatch(tree):
        raise McpError("ref did not resolve to immutable Git ids")
    return commit, tree


def json_at_ref(root: Path, ref: str, path: str) -> dict[str, Any]:
    process = subprocess.run(
        ["git", "-C", str(root), "show", f"{ref}:{path}"],
        capture_output=True,
        text=True,
        check=False,
    )
    if process.returncode != 0:
        raise McpError(f"{ref} has no {path}")
    try:
        value = json.loads(process.stdout)
    except json.JSONDecodeError as exc:
        raise McpError(f"{ref}:{path} is not JSON: {exc}") from exc
    if not isinstance(value, dict):
        raise McpError(f"{ref}:{path} must be an object")
    return value


def tool_name(command: dict[str, Any]) -> str:
    return f"loopctl_{command['loop']}_{command['mode']}"


def selected_modules(lock: dict[str, Any]) -> set[str]:
    modules = lock.get("modules")
    if not isinstance(modules, list):
        raise McpError("composition lock modules are malformed")
    return {item["id"] for item in modules if isinstance(item, dict) and isinstance(item.get("id"), str)}


def load_modules_at_ref(root: Path, ref: str, selected: set[str]) -> dict[str, dict[str, Any]]:
    modules: dict[str, dict[str, Any]] = {}
    for module_id in sorted(selected):
        modules[module_id] = json_at_ref(
            root, ref, f".arena/modules/{module_id}/module.json"
        )
    return modules


def validate_external_policy(
    contract: dict[str, Any],
    policy: dict[str, Any],
    lock: dict[str, Any],
    modules: dict[str, dict[str, Any]],
) -> list[dict[str, Any]]:
    sys.path.insert(0, str(HERE))
    import mcp_tools  # noqa: PLC0415

    tools = mcp_tools.build(contract, policy)
    selected = selected_modules(lock)
    for tool in tools:
        entry = tool["_policy"]
        module_id = entry["module"]
        if module_id not in selected:
            raise McpError(f"MCP policy module is not selected: {module_id}")
        module = modules.get(module_id)
        if module is None:
            raise McpError(f"MCP policy module manifest is absent: {module_id}")
        external = module.get("external_policy") or {}
        if external.get("exposed") is not True:
            raise McpError(f"module does not permit external exposure: {module_id}")
        loop_id = tool["_argv"]["loop"]
        loop_matches = [
            loop
            for loop in module.get("loops", [])
            if isinstance(loop, dict) and loop.get("id") == loop_id
        ]
        if len(loop_matches) != 1 or loop_matches[0].get("external_policy") != "allowlisted":
            raise McpError(
                f"tool {tool['name']} is not an allowlisted loop of module {module_id}"
            )
        if entry["mutation"] != external.get("mutation") and not (
            entry["mutation"] == "none" and external.get("mutation") == "disposable-worktree"
        ):
            raise McpError(f"tool mutation exceeds module policy: {tool['name']}")
        if entry["secrets"] != "none":
            raise McpError(
                f"broker-only secret delivery is not implemented for external tool {tool['name']}"
            )
    return tools


def module_closure(
    module_id: str,
    modules: dict[str, dict[str, Any]],
) -> list[str]:
    providers: dict[str, str] = {}
    for candidate, module in modules.items():
        for capability in module.get("provides", []):
            if capability in providers and providers[capability] != candidate:
                raise McpError(f"duplicate capability provider: {capability}")
            providers[capability] = candidate
    selected: set[str] = set()
    queue = [module_id]
    while queue:
        current = queue.pop(0)
        if current in selected:
            continue
        selected.add(current)
        for capability in modules[current].get("requires", []):
            if capability.startswith("external:"):
                continue
            provider = providers.get(capability)
            if provider is None:
                raise McpError(f"{current} has no provider for {capability}")
            queue.append(provider)
    return sorted(selected)


def normalize_prefix(value: str) -> str:
    path = Path(value)
    if path.is_absolute() or ".." in path.parts:
        raise McpError(f"module path is not portable: {value}")
    return path.as_posix().rstrip("/")


def closure_prefixes(
    closure: list[str],
    modules: dict[str, dict[str, Any]],
) -> list[str]:
    prefixes: set[str] = set()
    for module_id in closure:
        module = modules[module_id]
        for value in module.get("roots", []):
            prefixes.add(normalize_prefix(value))
        for component in (module.get("components") or {}).values():
            for value in component.get("paths", []):
                prefixes.add(normalize_prefix(value))
        prefixes.add(f".arena/modules/{module_id}/module.json")
    # These are immutable generated inputs, not host secrets.  They are needed
    # only when a selected module explicitly refers to them.
    prefixes.update(
        {
            ".arena/mcp-policy.json",
            ".arena/locks/bettor-arena.lock.json",
            ".arena/contexts.lock.json",
            "data/module-proof/subjects.lock.json",
            "data/context-capsules/driver-parity.json",
        }
    )
    return sorted(prefix for prefix in prefixes if prefix and prefix != ".")


def path_matches(path: str, prefix: str) -> bool:
    return path == prefix or path.startswith(prefix + "/")


def prune_worktree(worktree: Path, prefixes: list[str]) -> tuple[int, int]:
    process = subprocess.run(
        ["git", "-C", str(worktree), "ls-files", "-z"],
        capture_output=True,
        check=False,
    )
    if process.returncode != 0:
        raise McpError("cannot list disposable worktree files")
    kept = removed = 0
    for raw in process.stdout.split(b"\0"):
        if not raw:
            continue
        path = raw.decode("utf-8", errors="strict")
        if any(path_matches(path, prefix) for prefix in prefixes):
            kept += 1
            continue
        target = worktree / path
        if target.is_symlink() or target.is_file():
            target.unlink()
        removed += 1
    for directory in sorted(
        (path for path in worktree.rglob("*") if path.is_dir()),
        key=lambda item: len(item.parts),
        reverse=True,
    ):
        if directory.name == ".git":
            continue
        try:
            directory.rmdir()
        except OSError:
            pass
    return kept, removed


@contextmanager
def disposable_worktree(root: Path, commit: str) -> Iterator[tuple[Path, Path]]:
    base = Path(tempfile.mkdtemp(prefix="loopctl-mcp-"))
    worktree = base / "repo"
    try:
        process = subprocess.run(
            ["git", "-C", str(root), "worktree", "add", "--detach", str(worktree), commit],
            capture_output=True,
            text=True,
            check=False,
        )
        if process.returncode != 0:
            raise McpError(process.stderr.strip() or "git worktree add failed")
        yield base, worktree
    finally:
        if worktree.exists():
            subprocess.run(
                ["git", "-C", str(root), "worktree", "remove", "--force", str(worktree)],
                capture_output=True,
                check=False,
            )
        shutil.rmtree(base, ignore_errors=True)


def to_argv(tool: dict[str, Any], arguments: dict[str, Any]) -> list[str]:
    spec = tool["_argv"]
    allowed = {
        flag.lstrip("-").replace("-", "_"): flag for flag in spec["flags"]
    }
    unknown = sorted(set(arguments) - set(allowed))
    if unknown:
        raise McpError(f"undeclared argument(s): {unknown}")
    argv = [spec["loop"], spec["mode"]]
    for key, value in sorted(arguments.items()):
        flag = allowed[key]
        if isinstance(value, bool):
            if value:
                argv.append(flag)
            continue
        if not isinstance(value, str):
            raise McpError(f"argument {key} must be a string or boolean")
        candidate = Path(value)
        if candidate.is_absolute() or ".." in candidate.parts:
            raise McpError(f"server-host path is forbidden: {key}")
        argv.extend([flag, value])
    if "--json" not in argv:
        argv.append("--json")
    return argv


def safe_artifact_ref(value: object) -> str:
    if not isinstance(value, str) or not SAFE_ARTIFACT.fullmatch(value):
        raise McpError(f"unsafe artifact_ref: {value!r}")
    path = Path(value)
    if path.is_absolute() or ".." in path.parts:
        raise McpError(f"artifact_ref escapes bundle: {value!r}")
    return value


def materialize_inline_bundle(
    base: Path,
    arguments: dict[str, Any],
    max_bytes: int,
) -> Path:
    if set(arguments) != {"bundle"} or not isinstance(arguments["bundle"], dict):
        raise McpError("typed inline carrier accepts exactly one bundle object")
    bundle = arguments["bundle"]
    if set(bundle) != {"packet_ref", "files"} or not isinstance(bundle["files"], list):
        raise McpError("bundle must contain packet_ref and files")
    if not bundle["files"]:
        raise McpError("bundle.files must be non-empty")
    packet_ref = safe_artifact_ref(bundle["packet_ref"])
    target = base / "input"
    seen: set[str] = set()
    total = 0
    for index, item in enumerate(bundle["files"]):
        if not isinstance(item, dict) or set(item) != {
            "artifact_ref",
            "sha256",
            "content_base64",
        }:
            raise McpError(f"bundle.files[{index}] is not closed")
        artifact_ref = safe_artifact_ref(item["artifact_ref"])
        if artifact_ref in seen:
            raise McpError(f"duplicate artifact_ref: {artifact_ref}")
        seen.add(artifact_ref)
        try:
            content = base64.b64decode(item["content_base64"], validate=True)
        except (TypeError, ValueError) as exc:
            raise McpError(f"bundle.files[{index}] has invalid base64") from exc
        total += len(content)
        if total > max_bytes:
            raise McpError("decoded inline request exceeds policy limit")
        if hashlib.sha256(content).hexdigest() != item["sha256"]:
            raise McpError(f"inline artifact digest mismatch: {artifact_ref}")
        destination = target / artifact_ref
        destination.parent.mkdir(parents=True, exist_ok=True)
        destination.write_bytes(content)
    if packet_ref not in seen:
        raise McpError("packet_ref is not present in files")
    return target / packet_ref


def bounded_payload(process: subprocess.CompletedProcess[str], limit: int) -> dict[str, Any]:
    total = len(process.stdout.encode()) + len(process.stderr.encode())
    if total > limit:
        raise McpError("loopctl output exceeds policy limit")
    try:
        value = json.loads(process.stdout)
    except json.JSONDecodeError:
        return {
            "error": "loopctl produced no JSON result",
            "exit": process.returncode,
            "stdout": process.stdout[-4000:],
            "stderr": process.stderr[-4000:],
        }
    if not isinstance(value, dict):
        raise McpError("loopctl JSON result must be an object")
    return value


def execute_tool(
    root: Path,
    commit: str,
    tree: str,
    tool: dict[str, Any],
    modules: dict[str, dict[str, Any]],
    policy_digest: str,
    arguments: dict[str, Any],
) -> dict[str, Any]:
    policy = tool["_policy"]
    closure = module_closure(policy["module"], modules)
    prefixes = closure_prefixes(closure, modules)
    payload: dict[str, Any]
    cleanup = "FAIL"
    with disposable_worktree(root, commit) as (base, worktree):
        kept, removed = prune_worktree(worktree, prefixes)
        if not (worktree / "loopctl" / "loopctl.sh").is_file():
            raise McpError("selected module closure omitted loopctl")
        carrier = tool.get("_carrier")
        if carrier:
            if carrier.get("kind") != "ctg-inline-bundle@1.0.0":
                raise McpError(f"unsupported carrier: {carrier.get('kind')}")
            packet = materialize_inline_bundle(
                base, arguments, policy["max_request_bytes"]
            )
            output = base / "output"
            argv = [
                "ctg",
                "run",
                "--packet",
                str(packet),
                "--output",
                str(output),
                "--json",
            ]
        else:
            request_size = len(canonical(arguments))
            if request_size > policy["max_request_bytes"]:
                raise McpError("request exceeds policy limit")
            argv = to_argv(tool, arguments)
        environment = dict(os.environ)
        if policy["secrets"] == "none":
            for name in SECRET_NAMES:
                environment.pop(name, None)
        try:
            process = subprocess.run(
                ["sh", "loopctl/loopctl.sh", *argv],
                cwd=worktree,
                env=environment,
                capture_output=True,
                text=True,
                timeout=policy["max_seconds"],
                check=False,
            )
        except subprocess.TimeoutExpired as exc:
            raise McpError(f"tool timed out after {policy['max_seconds']} seconds") from exc
        payload = bounded_payload(process, policy["max_output_bytes"])
        payload["mcp_subject"] = {
            "module": policy["module"],
            "module_closure": closure,
            "commit": commit,
            "tree": tree,
            "policy_sha256": policy_digest,
            "kept_tracked_files": kept,
            "removed_tracked_files": removed,
            "owner_dependency_borrowing": False,
        }
    cleanup = "PASS"
    payload["mcp_subject"]["cleanup"] = cleanup
    return payload


def public_tool(tool: dict[str, Any]) -> dict[str, Any]:
    return {key: value for key, value in tool.items() if not key.startswith("_")}


def error(request_id: Any, code: int, message: str) -> dict[str, Any]:
    return {"jsonrpc": "2.0", "id": request_id, "error": {"code": code, "message": message}}


def result(request_id: Any, value: dict[str, Any]) -> dict[str, Any]:
    return {"jsonrpc": "2.0", "id": request_id, "result": value}


def load_surface(root: Path, commit: str) -> tuple[list[dict[str, Any]], dict[str, dict[str, Any]], str]:
    contract = json_at_ref(root, commit, "loopctl/contract.json")
    policy = json_at_ref(root, commit, ".arena/mcp-policy.json")
    lock = json_at_ref(root, commit, ".arena/locks/bettor-arena.lock.json")
    selected = selected_modules(lock)
    modules = load_modules_at_ref(root, commit, selected)
    tools = validate_external_policy(contract, policy, lock, modules)
    return tools, modules, digest_value(policy)


def handle(
    request: dict[str, Any],
    root: Path,
    commit: str,
    tree: str,
    tools: list[dict[str, Any]],
    modules: dict[str, dict[str, Any]],
    policy_digest: str,
) -> dict[str, Any] | None:
    method = request.get("method")
    request_id = request.get("id")
    if method == "initialize":
        return result(
            request_id,
            {
                "protocolVersion": PROTOCOL,
                "capabilities": {"tools": {}},
                "serverInfo": {"name": "loopctl", "version": commit},
            },
        )
    if method == "tools/list":
        return result(request_id, {"tools": [public_tool(tool) for tool in tools]})
    if method == "tools/call":
        params = request.get("params") or {}
        tool = next((item for item in tools if item["name"] == params.get("name")), None)
        if tool is None:
            return error(request_id, -32602, f"unknown or unexposed tool {params.get('name')!r}")
        arguments = params.get("arguments") or {}
        if not isinstance(arguments, dict):
            return error(request_id, -32602, "tool arguments must be an object")
        try:
            payload = execute_tool(
                root, commit, tree, tool, modules, policy_digest, arguments
            )
        except (McpError, OSError, subprocess.SubprocessError) as exc:
            payload = {"error": str(exc), "exit": 64}
        return result(
            request_id,
            {
                "content": [
                    {
                        "type": "text",
                        "text": json.dumps(payload, indent=2, ensure_ascii=False),
                    }
                ],
                "isError": payload.get("exit", 0) != 0,
            },
        )
    if method == "notifications/initialized":
        return None
    return error(request_id, -32601, f"unsupported method: {method}")


def parse_args(argv: list[str], environment: dict[str, str] | None = None) -> str:
    environment = os.environ if environment is None else environment
    ref = environment.get("LOOPCTL_REF", "")
    rest = list(argv)
    while rest:
        flag = rest.pop(0)
        if flag != "--ref" or not rest:
            raise McpError("usage: mcp_server.py --ref <40-hex|v-tag>")
        ref = rest.pop(0)
    if not ref:
        raise McpError("an immutable --ref or LOOPCTL_REF is required")
    return ref


def serve(ref: str) -> int:
    root = repo_root()
    commit, tree = resolve_ref(root, ref)
    tools, modules, policy_digest = load_surface(root, commit)
    for line in sys.stdin:
        try:
            request = json.loads(line)
            if not isinstance(request, dict):
                raise json.JSONDecodeError("request is not an object", line, 0)
        except json.JSONDecodeError as exc:
            print(json.dumps(error(None, -32700, f"parse error: {exc}")), flush=True)
            continue
        response = handle(request, root, commit, tree, tools, modules, policy_digest)
        if response is not None:
            print(json.dumps(response, ensure_ascii=False), flush=True)
    return 0


def _selftest() -> int:
    red = 0

    def case(name: str, got: Any, want: Any) -> None:
        nonlocal red
        if got != want:
            print(f"SELFTEST case failed — {name}: got {got!r}, want {want!r}", file=sys.stderr)
            red = 1

    root = repo_root()
    commit = git(root, "rev-parse", "HEAD")
    tree = git(root, "rev-parse", f"{commit}^{{tree}}")
    tools, modules, policy_digest = load_surface(root, commit)
    case("policy-exposes-a-bounded-set", 0 < len(tools) < 20, True)
    case("policy-digest-is-stable", len(policy_digest), 64)
    case("all-tools-name-a-module", all(tool["_policy"]["module"] in modules for tool in tools), True)
    try:
        parse_args(["--ref", "HEAD"], {})
        case("mutable-head-is-refused", "accepted", "McpError")
    except McpError:
        pass
    try:
        to_argv(tools[0], {"force_receipt": "/tmp/escape"})
        case("absolute-server-path-is-refused", "accepted", "McpError")
    except McpError:
        pass
    response = handle(
        {"jsonrpc": "2.0", "id": 1, "method": "tools/list"},
        root,
        commit,
        tree,
        tools,
        modules,
        policy_digest,
    )
    case("tools-list-hides-policy", all("_policy" not in tool for tool in response["result"]["tools"]), True)
    response = handle(
        {"jsonrpc": "2.0", "id": 2, "method": "tools/call", "params": {"name": "loopctl_macro_run"}},
        root,
        commit,
        tree,
        tools,
        modules,
        policy_digest,
    )
    case("unexposed-tool-is-unknown", "error" in response, True)
    source = Path(__file__).read_text(encoding="utf-8")
    case("no-owner-node-modules-borrowing", "symlink_to(root / factory)" in source, False)

    parent: Path | None = None
    with disposable_worktree(root, commit) as (_, worktree):
        parent = worktree.parent
        closure = module_closure(tools[0]["_policy"]["module"], modules)
        prune_worktree(worktree, closure_prefixes(closure, modules))
        case("unselected-notebooklm-is-absent", (worktree / "notebooklm").exists(), tools[0]["_policy"]["module"] == "notebooklm")
    case("disposable-worktree-is-cleaned", parent is not None and not parent.exists(), True)

    print("SELFTEST " + ("GREEN" if not red else "RED"))
    return red


def main(argv: list[str] | None = None) -> int:
    argv = sys.argv[1:] if argv is None else argv
    if argv[:1] == ["--selftest"]:
        return _selftest()
    try:
        ref = parse_args(argv)
        return serve(ref)
    except McpError as exc:
        print(f"mcp-server FATAL: {exc}", file=sys.stderr)
        return 64


if __name__ == "__main__":
    raise SystemExit(main())

File: loopctl/mcp_runtime.ts
#!/usr/bin/env bun
import { spawnSync } from "node:child_process";
import { existsSync } from "node:fs";
import { createServer } from "node:http";
import { createInterface } from "node:readline";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import {
  McpError,
  PROTOCOL,
  attachInlineDelivery,
  boundedJsonPayload,
  canonical,
  closurePrefixes,
  createWorkspace,
  loadSurface,
  moduleClosure,
  prepareInlineCarrier,
  pruneWorktree,
  publicTool,
  resolveRef,
  sanitizedEnvironment,
  toArgv,
  type GeneratedTool,
  type ModuleManifest,
  type PreparedInlineCarrier,
} from "./mcp_core.ts";

const HERE = dirname(fileURLToPath(import.meta.url));
const ROOT = resolve(HERE, "..");

interface RpcRequest {
  jsonrpc?: string;
  id?: string | number | null;
  method?: string;
  params?: { name?: string; arguments?: unknown };
}

interface ParsedArgs {
  ref: string;
  httpPort?: number;
}

function rpcError(id: RpcRequest["id"], code: number, message: string): object {
  return { jsonrpc: "2.0", id: id ?? null, error: { code, message } };
}

export function parseArgs(argv: string[], environment: Record<string, string | undefined> = process.env): ParsedArgs {
  let ref = environment.LOOPCTL_REF ?? "";
  let httpPort: number | undefined;
  const rest = [...argv];
  while (rest.length) {
    const flag = rest.shift();
    if (flag === "--ref") {
      const value = rest.shift();
      if (!value) {
        throw new McpError("--ref requires an immutable commit or v* tag");
      }
      ref = value;
    } else if (flag === "--http") {
      const value = rest.shift();
      if (!value || !/^\d+$/.test(value)) {
        throw new McpError("--http requires a numeric port");
      }
      httpPort = Number(value);
      if (httpPort < 1 || httpPort > 65535) {
        throw new McpError("--http port is outside 1..65535");
      }
    } else if (flag === "--json") {
      // loopctl declares --json on its own surface; server output is always structured.
    } else {
      throw new McpError(`unknown argument: ${flag}`);
    }
  }
  if (!ref) {
    throw new McpError("an immutable --ref or LOOPCTL_REF is required");
  }
  return httpPort === undefined ? { ref } : { ref, httpPort };
}

export function executeTool(
  root: string,
  commit: string,
  tree: string,
  tool: GeneratedTool,
  modules: Map<string, ModuleManifest>,
  policyDigest: string,
  argumentsValue: unknown,
): Record<string, unknown> {
  const policy = tool._policy;
  const closure = moduleClosure(policy.module, modules);
  const prefixes = closurePrefixes(closure, modules);
  const workspace = createWorkspace(root, commit);
  let payload: Record<string, unknown> | undefined;
  try {
    const pruned = pruneWorktree(workspace.worktree, prefixes);
    const loopctl = join(workspace.worktree, "loopctl/loopctl.sh");
    if (!pruned.kept || !existsSync(loopctl)) {
      throw new McpError("selected module closure omitted loopctl");
    }

    let argv: string[];
    let preparedCarrier: PreparedInlineCarrier | undefined;
    if (tool._carrier) {
      preparedCarrier = prepareInlineCarrier(tool, workspace.base, argumentsValue, policy.max_request_bytes);
      argv = preparedCarrier.argv;
    } else {
      if (Buffer.byteLength(canonical(argumentsValue ?? {})) > policy.max_request_bytes) {
        throw new McpError("request exceeds policy limit");
      }
      argv = toArgv(tool, argumentsValue ?? {});
    }

    const process = spawnSync("sh", ["loopctl/loopctl.sh", ...argv], {
      cwd: workspace.worktree,
      env: sanitizedEnvironment(),
      encoding: "utf8",
      timeout: policy.max_seconds * 1000,
      maxBuffer: policy.max_output_bytes + 64 * 1024,
    });
    if (process.error) {
      if ((process.error as NodeJS.ErrnoException).code === "ETIMEDOUT") {
        throw new McpError(`tool timed out after ${policy.max_seconds} seconds`);
      }
      throw new McpError(`tool process failed: ${process.error.message}`);
    }
    payload = boundedJsonPayload(
      process.stdout || "",
      process.stderr || "",
      process.status ?? 64,
      policy.max_output_bytes,
    );
    if (typeof payload.exit !== "number") {
      payload.exit = process.status ?? 64;
    }
    delete payload.stdout;
    delete payload.stderr;
    if (preparedCarrier) {
      attachInlineDelivery(payload, tool, preparedCarrier, policy.max_output_bytes);
    }
    payload.mcp_subject = {
      module: policy.module,
      module_closure: closure,
      commit,
      tree,
      policy_sha256: policyDigest,
      kept_tracked_files: pruned.kept,
      removed_tracked_files: pruned.removed,
      owner_dependency_borrowing: false,
      runtime: "bun-typescript",
    };
  } finally {
    workspace.cleanup();
  }
  if (!payload) throw new McpError("tool produced no payload");
  (payload.mcp_subject as Record<string, unknown>).cleanup = "PASS";
  return payload;
}

export function handle(
  request: RpcRequest,
  root: string,
  commit: string,
  tree: string,
  tools: GeneratedTool[],
  modules: Map<string, ModuleManifest>,
  policyDigest: string,
): object | undefined {
  if (request.method === "initialize") {
    return {
      jsonrpc: "2.0",
      id: request.id ?? null,
      result: {
        protocolVersion: PROTOCOL,
        capabilities: { tools: {} },
        serverInfo: { name: "loopctl", version: commit },
      },
    };
  }
  if (request.method === "tools/list") {
    return {
      jsonrpc: "2.0",
      id: request.id ?? null,
      result: { tools: tools.map(publicTool) },
    };
  }
  if (request.method === "tools/call") {
    const tool = tools.find((candidate) => candidate.name === request.params?.name);
    if (!tool) {
      return rpcError(request.id, -32602, `unknown or externally denied tool ${JSON.stringify(request.params?.name)}`);
    }
    try {
      const payload = executeTool(root, commit, tree, tool, modules, policyDigest, request.params?.arguments ?? {});
      return {
        jsonrpc: "2.0",
        id: request.id ?? null,
        result: {
          content: [{ type: "text", text: JSON.stringify(payload, null, 2) }],
          isError: payload.exit !== 0,
        },
      };
    } catch (error) {
      const payload = {
        error: String(error instanceof Error ? error.message : error),
        exit: 64,
      };
      return {
        jsonrpc: "2.0",
        id: request.id ?? null,
        result: {
          content: [{ type: "text", text: JSON.stringify(payload, null, 2) }],
          isError: true,
        },
      };
    }
  }
  if (request.method?.startsWith("notifications/")) return undefined;
  return rpcError(request.id, -32601, `unsupported method: ${request.method}`);
}

async function serveStdio(
  root: string,
  commit: string,
  tree: string,
  tools: GeneratedTool[],
  modules: Map<string, ModuleManifest>,
  policyDigest: string,
): Promise<number> {
  const lines = createInterface({ input: process.stdin, crlfDelay: Infinity });
  for await (const line of lines) {
    if (!line.trim()) continue;
    if (Buffer.byteLength(line) > 1024 * 1024) {
      console.log(JSON.stringify(rpcError(null, -32602, "request exceeds 1 MiB")));
      continue;
    }
    try {
      const request = JSON.parse(line) as RpcRequest;
      const response = handle(request, root, commit, tree, tools, modules, policyDigest);
      if (response) console.log(JSON.stringify(response));
    } catch (error) {
      console.log(JSON.stringify(rpcError(null, -32700, `parse error: ${String(error)}`)));
    }
  }
  return 0;
}

async function serveHttp(
  root: string,
  commit: string,
  tree: string,
  tools: GeneratedTool[],
  modules: Map<string, ModuleManifest>,
  policyDigest: string,
  port: number,
): Promise<number> {
  const server = createServer(async (request, response) => {
    if (request.method !== "POST" || request.url?.replace(/\/$/, "") !== "/mcp") {
      response.writeHead(404).end();
      return;
    }
    const chunks: Buffer[] = [];
    let total = 0;
    for await (const chunk of request) {
      const buffer = Buffer.from(chunk);
      total += buffer.length;
      if (total > 1024 * 1024) {
        response.writeHead(413, { "content-type": "application/json" });
        response.end(JSON.stringify(rpcError(null, -32602, "request exceeds 1 MiB")));
        return;
      }
      chunks.push(buffer);
    }
    try {
      const rpc = JSON.parse(Buffer.concat(chunks).toString("utf8")) as RpcRequest;
      const value = handle(rpc, root, commit, tree, tools, modules, policyDigest);
      if (!value) {
        response.writeHead(202).end();
        return;
      }
      const body = JSON.stringify(value);
      response.writeHead(200, {
        "content-type": "application/json",
        "content-length": Buffer.byteLength(body),
      });
      response.end(body);
    } catch (error) {
      const body = JSON.stringify(rpcError(null, -32700, `parse error: ${String(error)}`));
      response.writeHead(200, { "content-type": "application/json" });
      response.end(body);
    }
  });
  await new Promise<void>((accept, reject) => {
    server.once("error", reject);
    server.listen(port, "127.0.0.1", accept);
  });
  console.error(`mcp: serving ref=${commit} on http://127.0.0.1:${port}/mcp`);
  await new Promise<void>((accept) => {
    process.once("SIGINT", () => server.close(() => accept()));
    process.once("SIGTERM", () => server.close(() => accept()));
  });
  return 0;
}

function selftest(): number {
  let red = 0;
  const expectError = (name: string, operation: () => unknown): void => {
    try {
      operation();
      console.error(`SELFTEST case failed — ${name}`);
      red = 1;
    } catch {
      // expected
    }
  };
  expectError("missing-ref", () => parseArgs([], {}));
  expectError("mutable-head", () => resolveRef(ROOT, "HEAD"));
  const init = handle(
    { jsonrpc: "2.0", id: 1, method: "initialize" },
    ROOT,
    "a".repeat(40),
    "b".repeat(40),
    [],
    new Map(),
    "c".repeat(64),
  ) as { result?: { protocolVersion?: string } };
  if (init.result?.protocolVersion !== PROTOCOL) red = 1;
  const unknown = handle(
    {
      jsonrpc: "2.0",
      id: 2,
      method: "tools/call",
      params: { name: "nope" },
    },
    ROOT,
    "a".repeat(40),
    "b".repeat(40),
    [],
    new Map(),
    "c".repeat(64),
  ) as { error?: { message?: string } };
  if (!unknown.error?.message?.includes("externally denied")) red = 1;
  console.log(`SELFTEST ${red ? "RED" : "GREEN"}`);
  return red;
}

export async function main(argv: string[]): Promise<number> {
  if (argv[0] === "--selftest") return selftest();
  let parsed: ParsedArgs;
  try {
    parsed = parseArgs(argv);
  } catch (error) {
    console.error(`mcp-server FATAL: ${String(error instanceof Error ? error.message : error)}`);
    return 64;
  }
  try {
    const ids = resolveRef(ROOT, parsed.ref);
    const surface = loadSurface(ROOT, ids.commit);
    return parsed.httpPort
      ? serveHttp(ROOT, ids.commit, ids.tree, surface.tools, surface.modules, surface.policyDigest, parsed.httpPort)
      : serveStdio(ROOT, ids.commit, ids.tree, surface.tools, surface.modules, surface.policyDigest);
  } catch (error) {
    console.error(`mcp-server FATAL: ${String(error instanceof Error ? error.message : error)}`);
    return 64;
  }
}

const invoked = process.argv[1] && resolve(process.argv[1]) === fileURLToPath(import.meta.url);
if (invoked) process.exit(await main(process.argv.slice(2)));
